_posts/2019-04-16-preservation.md

```diff
@@ -0,0 +1,358 @@
+
+1. We have a bag that people can look at and reasonably understand.
+1. External systems may possible be able to ingest the material, but in practice
+   this is pretty unlikely - bag profiles vary substantially from vendor to
+   vendor.
+
+**Cons**:
+
+1. The manifest files act as a pessimistic lock on the system.
+   * If we have to wait until everything is "done" uploading to update those
+      manifests, then we need to have some way to tell when we're "done." To do
+      this, in Ruby at least, either the uploads must be done sequentially on a
+      single system or some sort of external locking mechanism must be used to
+      keep track of current upload state.
+   * If instead of waiting we simply update the manifest when a file is
+      uploaded, then it means we need to have the ability to lock the file itself
+      in place to avoid writes from multiple processes running into each other.
+1. We have to prevent future changes to a child from persisting to preservation
+   storage until this large change is done. Otherwise, the manifest will be
+   corrupted by a race condition:
+   1. Preservation starts. Pages start to persist into a bag.
+   2. Page gets modified out of band (label changed or page switched)
+   3. Page's changes get sent to be preserved.
+   4. Page gets preserved in its spot in the bag, updating manifest.
+   5. Initial preservation ends. Updates manifest, but has out-of-date checksum
+      information.
+
+#### OCFL
+
+Implementation notes for OCFL recommend things such as pair-trees for storing
+objects in the object root. I'm going to treat that as an implementation detail
+and pretend we're using a file system that can handle an infinite number of
+resources in the same directory.
+
+1. Create a base directory named `<book-id>`
+1. Create a file named `0=ocfl_object_1.0` with the appropriate contents.
+1. Create a v1 directory
+1. Copy all content files and metadata files into `<book-id>/v1/content`, separated
+   by ID. So, for instance, `<page1>` would be in `<book-id>/v1/content/<page1>`,
+   and have two files: `<page1>.json` and
+   `<page1>-<file-name>.<file-extension>`.
+1. Create an `inventory.json` in `v1`, as per the specification, with SHA512s of
+   every file in `content` as the keys of the `manifest` property.
+1. Create `inventory.json.sha512` in the `v1` directory.
+1. Copy `inventory.json` and `inventory.json.sha512` to the `<book-id>`
+   directory.
+
+**Pros**:
+
+1. We now have an OCFL repository which is theoretically readable by future OCFL
+   browsers, and will be understood by partner institutions.
+1. This structure just needs a file system to store itself and support
+   versioning, there's no need for an object store's feature-set.
+1. We can de-duplicate as part of generating the `inventory.json` if we'd like
+   to, using up less space than the other options.
+
+**Cons**:
+
+1. The inventory file acts as a pessimistic lock in the same way as BagIt's
+   manifest file.
```

Comment:
Is this a problem you face with OCFL and BagIt or with any kind of versioning?

Reply:
OCFL and BagIt. The problem is there's a single file which needs to be updated to represent a logical unit made up of several resources. Every write to any resource results in having to write that same file. If we rely on GCS' versioning, or any system which doesn't have a single file inventory, then there's no point of contention and I can append pieces of the preservation store with no worry.

Reply:
I agree that OCFL and BagIt have essentially the same property with manifest/inventory in this regard. In both cases updates to an object should be made as an atomic operation -- locked from other operations on the same object -- but it is an implementation choice whether the manifest/inventory is kept up-to-date in the middle of this overall operation. Support for a resource-within-object centric interface would suggest updating the manifest/inventory for each change which I agree would be a bottleneck. This seems a tension between preservation vs active-data approaches